src/modules/job-exec/sdexec.c

```diff
@@ -191,6 +214,218 @@ static void set_stdlog (struct sdexec *se,
+
+    if (!idset)
+        return -1;
+
+    (*in_idset) = idset_test (idset, rank);
+    idset_destroy (idset);
+    return 0;
+}
+
+static char *hwloc_cpuset_2_systemd_property (hwloc_cpuset_t set)
+{
+    char *str = NULL;
+    size_t str_index = 0;
+    size_t str_len = 0;
+    int i;
+
+    /* N.B. It is assumed hwloc bitmap will always returns id in
+     * increasing numerical order */
+
+    i = hwloc_bitmap_first (set);
+    while (i >= 0) {
+        int len;
+        /* max printed int is 10 chars + comma = 11, we'll realloc if
+         * we're ever below 16 bytes left */
+        if ((str_len - str_index) < 16) {
+            str_len += 256;
+            if (!(str = realloc (str, str_len)))
+                goto error;
+        }
+        len = snprintf (str + str_index,
+                        str_len - str_index,
+                        "%s%d",
+                        str_index ? "," : "CPUAffinity=",
+                        i);
+        assert (len < (str_len - str_index));
+        str_index += len;
+        i = hwloc_bitmap_next (set, i);
+    }
+
+    return str;
+
+error:
+    free (str);
+    return NULL;
+}
+
+static char *CPUAffinity_list (struct jobinfo *job,
+                               const char *cores)
+{
+    hwloc_topology_t topo = NULL;
+    hwloc_cpuset_t coreset = NULL;
+    hwloc_cpuset_t puset = NULL;
+    int depth;
+    int i;
+    char *rv = NULL;
+
+    if (hwloc_topology_init (&topo) < 0
+        || hwloc_topology_load (topo) < 0) {
+        flux_log_error (job->h, "hwloc topology setup");
+        goto error;
+    }
+
+    depth = hwloc_get_type_depth (topo, HWLOC_OBJ_CORE);
+    if (depth == HWLOC_TYPE_DEPTH_UNKNOWN
+        || depth == HWLOC_TYPE_DEPTH_MULTIPLE) {
+        flux_log_error (job->h, "invalid depth returned");
+        goto error;
+    }
+
+    if (!(coreset = hwloc_bitmap_alloc ())
+        || !(puset = hwloc_bitmap_alloc ())) {
+        flux_log_error (job->h, "hwloc_bitmap_alloc");
+        goto error;
+    }
+
+    if (hwloc_bitmap_list_sscanf (coreset, cores) < 0) {
+        flux_log_error (job->h, "hwloc_bitmap_list_sscanf");
+        goto error;
+    }
+
+    /* A core may be hyperthreaded, so we have to figure out all
+     * processor units that need to be "affinitied" */
+
+    i = hwloc_bitmap_first (coreset);
+    while (i >= 0) {
+        hwloc_obj_t core;
+        int j;
+        if (!(core = hwloc_get_obj_by_depth (topo, depth, i))) {
+            flux_log_error (job->h, "hwloc_get_obj_by_depth");
+            goto error;
+        }
+        if (!core->arity || !core->children) {
+            flux_log_error (job->h, "hwloc obj invalid");
+            goto error;
+        }
+        for (j = 0; j < core->arity; j++) {
+            if (core->children[j]->type != HWLOC_OBJ_PU) {
+                flux_log_error (job->h, "child of core obj not a pu");
+                goto error;
+            }
+            hwloc_bitmap_set (puset, core->children[j]->os_index);
+        }
```

Comment:
This function is largely a copy of what is done in `src/shell/affinity.c` `shell_affinity_get_cpuset()`.
Perhaps not necessary to copy that function, nor to create a libutil function at this point, but a simplification from the existing implementation would be to use `hwloc_bitmap_or ()` on `core->cpuset` instead of by hand setting the bit of each core child. (This lets you drop this entire loop).